qa/L0_libtorch_shared_weights/libtorch_shared_weights_test.py

```diff
@@ -0,0 +1,93 @@
+# are met:
+#  * Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+#  * Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#  * Neither the name of NVIDIA CORPORATION nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import sys
+
+sys.path.append("../common")
+
+import argparse
+import numpy as np
+import requests as httpreq
+import unittest
+from builtins import range
+import tritonhttpclient as httpclient
+import test_util as tu
+
+FLAGS = None
+
+
+class SharedWeightsTest(tu.TestResultCollector):
+
+    def _full_exact(self, model_name, request_parallelism, shape):
+
+        # Run async requests to make sure backend handles parallel requests
+        # correctly.
+        client = httpclient.InferenceServerClient(
+            "localhost:8000", concurrency=request_parallelism)
+        input_datas = []
+        requests = []
+        for i in range(request_parallelism):
+            input_data = (16384 * np.random.randn(*shape)).astype(np.float32)
+            input_datas.append(input_data)
+            inputs = [
+                httpclient.InferInput("INPUT__0", input_data.shape, "FP32")
+            ]
+            inputs[0].set_data_from_numpy(input_data)
+            requests.append(client.async_infer(model_name, inputs))
+
+        for i in range(request_parallelism):
+            # Get the result from the initiated asynchronous inference request.
+            # Note the call will block until the server responds.
+            results = requests[i].get_result()
+
+            output_data = results.as_numpy("OUTPUT__0")
+            self.assertIsNotNone(output_data,
+                                 "error: expected 'OUTPUT__0' to be found")
+            np.testing.assert_allclose(output_data, input_datas[i])
+
+        # Make sure the requests ran in parallel.
```

Comment:
How are we testing that requests ran in parallel? With an instance count of 4, I presume only 4 requests will run at a time in pytorch.

Reply:
Good point. I read through the documentation and see now that requests for that model wait until an instance is free, if all the instances are taken. I fixed the comments. I also updated the instance count to 16 and concurrency to 128, since it's a small model (test takes a few seconds), to ensure everything works with greater parallelism.